spec-0007/transition_to_rng.py

```diff
@@ -0,0 +1,156 @@
+                f"of `{old_name}`. For example, "
+                "1) `None` results in unpredictable random numbers, "
+                "2) an integer results in a different stream of random numbers, "
+                "(with the same distribution), and "
+                "3) `np.random` or `RandomState` instances result in an error. "
+                "See the documentation of `default_rng` for more information."
+            )
+
+            if as_old_kwarg:  # warn about deprecated use of old kwarg
+                kwargs[NEW_NAME] = kwargs.pop(old_name)
+                if dep_version:
+                    message = (
+                        f"Use of keyword argument `{old_name}` is "
+                        f"deprecated and replaced by `{NEW_NAME}`.  "
+                        f"Support for `{old_name}` will be removed "
+                        f"in SciPy {end_version}."
+                    ) + cmn_msg
+                    warnings.warn(message, DeprecationWarning, stacklevel=2)
+
+            elif as_pos_arg and dep_version:
+                # Warn about changing meaning of positional arg
+
+                # Note that this decorator does not deprecate positional use of the
+                # argument; it only warns that the behavior will change in the future.
+                # Simultaneously transitioning to keyword-only use is another option.
+
+                message = (
+                    f"Positional use of `{NEW_NAME}` (formerly known as "
+                    f"`{old_name}`) is still allowed, but the behavior is "
+                    "changing: the argument will be validated using "
+                    f"`np.random.default_rng` beginning in SciPy {end_version}."
+                ) + cmn_msg
+                warnings.warn(message, FutureWarning, stacklevel=2)
+
+            elif as_new_kwarg:  # no warnings; this is the preferred use
+                # After the removal of the decorator, validation with
+                # np.random.default_rng will be done inside the decorated function
+                kwargs[NEW_NAME] = np.random.default_rng(kwargs[NEW_NAME])
+
+            elif np.random.mtrand._rand._bit_generator._seed_seq is None:
+                # Emit FutureWarning if `np.random.seed` was used and no PRNG was passed
+                message = (
+                    "The NumPy global RNG was seeded by calling "
+                    f"`np.random.seed`. Beginning in {end_version}, this "
+                    "function will no longer use the global RNG."
+                ) + cmn_msg
+                warnings.warn(message, FutureWarning, stacklevel=2)
+
+            return fun(*args, **kwargs)
+
+        return wrapper
+
+    return decorator
+
+
+# Example usage of _prepare_rng decorator
+@_transition_to_rng("random_state", 1, "1.15.0")
+# previously, the signature of the function was
+#   library_function(arg1, random_state=None, arg2=None):
+def library_function(arg1, rng=None, arg2=None):
```

Comment:
May want to use `*, rng=` here since that is the recommendation after all.

Reply:
This is showing how to transition from an old function in which `random_state` is not keyword-only, and the SPEC does not step so far as to recommend that the new argument should be keyword only. We had that, but took it out. So this is maintaining positional use intentionally.